### ISSUE TITLE
`invalid-attribute-override` - false positive when overriding `ClassVar` with "class or instance" attribute

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3992]
### Summary

```py
from typing import ClassVar

class Base:
    instance_attr: int
    class_attr: ClassVar[int]

class Sub(Base):
    class_attr: int  # error: [invalid-attribute-override]

Sub.class_attr = 1 # no error
```
[playground](https://play.ty.dev/2480e980-1584-445e-8249-3c58ad0c33bf)

see discussion in https://github.com/astral-sh/ty/issues/3721#issuecomment-4676128051.

to summarize, from my understanding there are three separate attribute types:

- class-only attributes (ie. `ClassVar`s)
- regular attributes (regular attributes defined on the class)
- instance-only attributes (defined on the instance)

since regular attributes can be used as either class attributes or instance attributes, they should be considered a "subtype" of both class-only and instance-only attributes, so it should not be an error to override a class-attribute with a regular attribute

### Version

bfa66e606

**Comment by sharkdp** [comment]
The idea is that ty understands a declaration like `class_attr: int` as being an instance-only attribute (your third bullet point). Given that interpretation, this seems like a clear Liskov violation (because `Base`'s `class_attr` annotation claims that `class_attr` is available on the class object itself). A regular attribute would be something like this, which does not lead to an error:

```py
class Sub(Base):
    class_attr = 1
```


That leaves the question how `class_attr: int = 1` should be handled. It's currently an error.

What's your concrete use case? Why is ty's behavior a problem for you? Why do you need to override the attribute at all?

**Comment by DetachHead** [comment]
> Given that interpretation, this seems like a clear Liskov violation (because `Base`'s `class_attr` annotation claims that `class_attr` is available on the class object itself).

i don't really understand exactly *how* it's a liskov violation though. the way i see it, a regular attribute is a subtype of a class attribute because it can do everything a class attribute can do, and more. an instance-only attribute on the other hand cannot be used as a class attribute.

> What's your concrete use case? Why is ty's behavior a problem for you? Why do you need to override the attribute at all?

honestly i don't remember lol. i'm trying to switch type checkers on a massive codebase and i've opened issues for most of the problems i've run into in the process, but i seem to have lost track of where this one came from.

it does seem to match what pyright (our current type checker) does, so it probably isn't that much of a problem for me

**Comment by sharkdp** [comment]
> > Given that interpretation, this seems like a clear Liskov violation (because `Base`'s `class_attr` annotation claims that `class_attr` is available on the class object itself).
> 
> i don't really understand exactly _how_ it's a liskov violation though. the way i see it, a regular attribute is a subtype of a class attribute because it can do everything a class attribute can do, and more.

I agree that a regular attribute should have that behavior. What I'm saying is that `class_attr: int` is currently not interpreted as a regular attribute, but as a pure instance attribute.

**Comment by DetachHead** [comment]
are you saying that ty doesn't differentiate between regular attributes and instance only attributes? it does seem to understand that `class_attr` is a regular attribute from what i can tell, because it's treated differently to an instance-only attribute. for example:

```py
from typing import ClassVar

class Base:
    instance_attr: int
    class_attr: ClassVar[int]

    def __init__(self):
        self.instance_only_attribute: int = 1

class Sub(Base):
    class_attr: int  # error: [invalid-attribute-override]


# no errors when accessing the attribute on the class or the instance, so ty is treating it as a "regular" attribute:
print(Sub.class_attr) # no error
print(Sub().class_attr) # no error

# error when accessing the attribute on the class, so ty treats it as an instance only attribute
print(Sub.instance_only_attribute) # error: unresolved-attribute
print(Sub().instance_only_attribute) # no error
```
[playground](https://play.ty.dev/77313482-4a62-4c0c-8b66-f3817f5d4163)

**Comment by sharkdp** [comment]
Hm, yes, thank you for pushing back. That seems inconsistent. And I was mistaken: `class_attr: int` is (now) treated as a regular attribute, available on both class and instance. So I agree that we shouldn't emit `invalid-attribute-override`.

It seems like our distinction in Liskov checking is too coarse.

cc @charliermarsh as the author of https://github.com/astral-sh/ruff/pull/24767

**Comment by KotlinIsland** [comment]
i think that it should be an error for an instance variable

`ClassVar` has nothing to do with LSP or types, it's a matter of finality. but finality that is scoped only to instances, classes can still override

but why can't you infer a `ClassVar` from the super type?

```py
class A:
    a: ClassVar = 1

class B(A):
    a = 2  # inferred as ClassVar
```
